local/src/core/http2.cc

```diff
@@ -731,35 +736,61 @@ on_frame_send_cb(nghttp2_session *session, nghttp2_frame const *frame, void *use
-                                stream_state._client_rst_stream_after;
-    auto const error_code = stream_state._client_rst_stream_error == -1 ?
-                                stream_state._server_rst_stream_error :
-                                stream_state._client_rst_stream_error;
-    if (frame->hd.type == frame_type) {
+    auto const rst_stream_after = stream_state._client_rst_stream_after == -1 ?
+                                      stream_state._server_rst_stream_after :
+                                      stream_state._client_rst_stream_after;
+    auto const rst_stream_error = stream_state._client_rst_stream_error == -1 ?
+                                      stream_state._server_rst_stream_error :
+                                      stream_state._client_rst_stream_error;
+    auto const goaway_after = stream_state._client_goaway_after == -1 ?
+                                  stream_state._server_goaway_after :
+                                  stream_state._client_goaway_after;
+    auto const goaway_error = stream_state._client_goaway_error == -1 ?
+                                  stream_state._server_goaway_error :
+                                  stream_state._client_goaway_error;
+    if (frame->hd.type == rst_stream_after) {
       errata.note(
           S_DIAG,
           "Submitting RST_STREAM frame for key {} after {} frame with error code {}.",
           stream_state._key,
-          H2FrameNames[static_cast<H2Frame>(frame_type)],
-          H2ErrorCodeNames[static_cast<H2ErrorCode>(error_code)]);
+          H2FrameNames[static_cast<H2Frame>(rst_stream_after)],
+          H2ErrorCodeNames[static_cast<H2ErrorCode>(rst_stream_error)]);
       errata.sink();
-      int const submit_status = nghttp2_submit_rst_stream(session, 0, stream_id, error_code);
+      int const submit_status = nghttp2_submit_rst_stream(session, 0, stream_id, rst_stream_error);
       if (submit_status) {
         errata.note(
             S_DIAG,
             "Failed to submit RST_STREAM frame for key {} after {} frame: {}.",
             stream_state._key,
-            H2FrameNames[static_cast<H2Frame>(frame_type)],
+            H2FrameNames[static_cast<H2Frame>(rst_stream_after)],
             submit_status);
+      } else {
+        errata.note(
+            S_DIAG,
+            "Sent RST_STREAM frame for key {} on stream {}.",
+            stream_state._key,
+            stream_id);
+      }
+    } else if (frame->hd.type == goaway_after) {
+      errata.note(
+          S_DIAG,
+          "Submitting GOAWAY frame for key {} after {} frame with error code {}.",
+          stream_state._key,
+          H2FrameNames[static_cast<H2Frame>(goaway_after)],
+          H2ErrorCodeNames[static_cast<H2ErrorCode>(goaway_error)]);
+      errata.sink();
+      int const submit_status = nghttp2_session_terminate_session2(session, 0, goaway_error);
+      if (submit_status) {
```

Comment:
Nit: can you make this `submit_status != 0`. Only with boolean vars do we not compare to some value.